src/posix/platform/resolver.cpp

```diff
@@ -0,0 +1,293 @@
+
+exit:
+    if (error != OT_ERROR_NONE)
+    {
+        otLogCritPlat("Failed to forward DNS query %p to server: %d", static_cast<void *>(aTxn), error);
+    }
+    return;
+}
+
+void Resolver::Cancel(otPlatDnsUpstreamQuery *aTxn)
+{
+    Transaction *txn = GetTransaction(aTxn);
+
+    CloseTransaction(txn);
+}
+
+Resolver::Transaction *Resolver::AllocateTransaction(otPlatDnsUpstreamQuery *aThreadTxn)
+{
+    int          i;
+    int          error     = 0;
+    int          fdOrError = 0;
+    Transaction *ret       = nullptr;
+
+    for (i = 0; i < kMaxUpstreamTransactionCount; i++)
+    {
+        if (mUpstreamTransaction[i].mThreadTxn == nullptr)
+        {
+            ret       = &mUpstreamTransaction[i];
+            fdOrError = socket(AF_INET, SOCK_DGRAM, 0);
+            if (fdOrError < 0)
+            {
+                otLogInfoPlat("Failed to create socket: %d", fdOrError);
+                error = fdOrError;
+                break;
+            }
+            ret->mUdpFd                        = fdOrError;
+            mUpstreamTransaction[i].mThreadTxn = aThreadTxn;
+            break;
+        }
+    }
+
+    if (ret != nullptr && error != 0)
+    {
+        CloseTransaction(ret);
+        ret = nullptr;
+    }
+
+    return ret;
+}
+
+void Resolver::ForwardResponse(Transaction *aTxn)
+{
+    char       response[kMaxDnsMessageSize];
+    ssize_t    readSize;
+    otMessage *message;
+
+    VerifyOrExit((readSize = read(aTxn->mUdpFd, response, sizeof(response))) > 0);
+
+    message = otUdpNewMessage(gInstance, nullptr);
+    SuccessOrExit(otMessageAppend(message, response, readSize));
+
+    otPlatDnsUpstreamQueryDone(gInstance, aTxn->mThreadTxn, message);
+    message = nullptr;
+
+exit:
+    if (message != nullptr)
+    {
+        otMessageFree(message);
+    }
+}
+
+Resolver::Transaction *Resolver::GetTransaction(int aFd)
+{
+    Transaction *ret = nullptr;
+
+    for (int i = 0; i < kMaxUpstreamTransactionCount; i++)
+    {
+        if (mUpstreamTransaction[i].mThreadTxn != nullptr && mUpstreamTransaction[i].mUdpFd == aFd)
+        {
+            ret = &mUpstreamTransaction[i];
+            break;
+        }
+    }
+
+    return ret;
+}
+
+Resolver::Transaction *Resolver::GetTransaction(otPlatDnsUpstreamQuery *aThreadTxn)
+{
+    Transaction *ret = nullptr;
+
+    for (int i = 0; i < kMaxUpstreamTransactionCount; i++)
+    {
+        if (mUpstreamTransaction[i].mThreadTxn == aThreadTxn)
+        {
+            ret = &mUpstreamTransaction[i];
+            break;
+        }
+    }
+
+    return ret;
+}
+
+void Resolver::CloseTransaction(Transaction *aTxn)
+{
+    if (aTxn->mUdpFd >= 0)
+    {
+        close(aTxn->mUdpFd);
+        aTxn->mUdpFd = -1;
+    }
+    aTxn->mThreadTxn = nullptr;
+}
+
+void Resolver::UpdateFdSet(fd_set *aReadFdSet, fd_set *aErrorFdSet, int *aMaxFd)
+{
+    for (int i = 0; i < kMaxUpstreamTransactionCount; i++)
```

Comment:
This is fine as is but can consider using range-based for to go over the array.
I know `clang-tidy` may add a warning about it(not sure if we run `clang-tidy` for posix platform though).